### ISSUE TITLE
Support JSON output files as part of execution

### ISSUE THREAD

**Issue by teaglebuilt** [robotframework/robotframework#3423]
Problem:

Prefer to use a NOSQL database to store test reports. NOSQL databases typically require json format because its a relational database.

Solution:

Provide robotframework with the ability to use json rather than xml in the output file. 

For example:

```
           # flag --output
          robot --output output.json ./tests
```

**Comment by teaglebuilt** [comment]
If its ok with the community, I would like to work on this. I created a pull request for tracking progress and questions along the way

**Comment by pekkaklarck** [comment]
Being able to use output.json instead of output.xml may be useful. Biggest possible benefits I see are:

- JSON is a lot less verbose than XML, so output.json file would be a lot smaller than output.xml.
- Because output.json would be smaller, writing it to disk during execution would be faster as well.
- It is possible that processing output.json would be faster than processing output.xml. This is not certain, though, as our XML processing code is pretty well optimized and the underlying XML modules are pretty fast.

If processing output.json is faster than processing output.xml, then this feature is definitely a good idea. If it isn't, then we need to think are the other benefits big enough. This is something we can only know after prototyping this a bit.

I don't thing possibility to add data to a NOSQL database or to any other system is a good reason to add anything like this. It is already now possible to do that by parsing the information from output.xml using the standard XML modules or the `robot.api.ExecutionResult` API. Data being stored in JSON instead of XML doesn't really make a big difference.

**Comment by pekkaklarck** [comment]
The are two design decisions to be made:

1. How to enable this feature. I don't think any new command line option is needed. We just need to make it possible to say `--output output.json` and detect the type automatically. Also Rebot needs to be enhanced to allow using `rebot output.json`.

2. The output.json format. Because it would be a new external interface, it needs to be designed carefully. No need to think about that too much in the beginning, though. It is more important to test are there performance benefits or not. If there are, then we need to design the format more carefully.

**Comment by mkorpela** [comment]
There is also another option to reduce output size: write directly to tar.gz.
I think both reading and writing of output.xml would be fairly easy to modify to slso being able to read and write (the same xml) to output.tar.gz

**Comment by mkorpela** [comment]
https://docs.python.org/3/library/tarfile.html
(Zip and others also could work)

**Comment by teaglebuilt** [comment]
@mkorpela one reason why i am set on an json option is the possibilities of using NoSQL databases to easily store robotframework results.

**Comment by pekkaklarck** [comment]
Can you @teaglebuilt explain how this feature would help storing results in a db?

**Comment by teaglebuilt** [comment]
@pekkaklarck 
JSON is the query language used in NOSQL databases.

Well, MongoDB for example, is technically BSON format which is relative to JSON. from what ive seen there are many complications to storing xml in a NOSQL database. These are some examples below, 

Mongodb
Cassandra
Couch DB

They all require xml to be converted into JSON for storage. Unless you want to just save as a string which would not be able to query

**Comment by teaglebuilt** [comment]
@pekkaklarck with the possibility of output.json, we could essentially use NoSQL databases to store and query results. Then look into other possibilities of filtering and searching specific queries relative to the data in the output file

**Comment by teaglebuilt** [comment]
STATUS UPDATE: working on JSONELEMENTHANDLER, rebot accepting endswith(.json) on output file....

**Comment by pekkaklarck** [comment]
1. I still don't understand how Robot producing json output itself helps using any db. It is already now easy to read information from output.xml and write it to a db (or to a json file). If you have a db that accepts Robot's output.json as-is, then you avoid one step between test run and storing info to the db. Is that your motivation?

2. Great that you have some progress?

3. Have you already thought about the json format?

**Comment by sdave2** [comment]
@teaglebuilt could you share the JSON format you're looking to produce?
I am also under a similar situation where I want to get the output as JSON format.

Parsing XML and converting that to JSON I found to be very quick and also the XML gets converted into python objects and then I can use it like `suite.suites` and so on.

So few things come to mind:

1. Can you share the JSON structure you're looking to output

2. Also, when I read the JSON output back into python for further processing, will it be converted into python objects?

I was looking at this for JSON structure and JSON to python objects:
1. https://json-schema.org/understanding-json-schema/
2. https://pypi.org/project/python-jsonschema-objects/0.0.13/

Also, please let me know if i can help.

**Comment by Lemonlemmings** [comment]
So I've taken a fork and created a [JSON Schema](https://github.com/Lemonlemmings/robotframework/blob/master/doc/schema/robot-10.schema.json) and added a "--json" option to rebot so you can convert your XML results into JSON.

There are a couple of things which I have done which are probably not consistent with robot. 

- In the JSON schema I have named what robot normally labels as "kw" as "keyword" because for me it's easier for other people to understand if it's spelt out fully.
- In situations where it seemed appropiate to use a boolean I used a boolean. This is rather than "yes"/"no" for marking stats as critical / marking stats as "html" or not.
- I have not fully broken down the objects in the schema. JSON schemas have the idea of inheritance, and quite a few of the objects share some of the same fields, but I thought they were encapsulated in a way which was easy to understand.
- My python implementation is a little lazy. I haven't written documentation on my functions, and have strayed away from putting individual comments on lines (which I normally do, but was worried that the coding guidelines for robotframework wouldn't allow that). I also haven't written any acceptance tests. I plan on doing this, if there is interest in merging it in.
- Most of my schema was derived from the XSD. I could be wrong, but based off of my interactions which robot objects, and with the XSD there is a mismatch between the two. I think my JSON schema corrects some of those wrongs, but if there's a difference (which to me seems likely) then the supporting code would need to be changed.

I appreciate that it's been said that this could be done as part of the robot output. I don't think it'd be hard to adapt what I've done to that. Having said that though, I figured that rather than attaching the JSON to the robot implementation it was better as rebot. If you want the format in something else (i.e: Xunit) you simply use rebot to fix it up.

I'm quite new to github, and haven't done any contribution before now. So there maybe some rules which I'm not so aware of, so if I'm doing something stupid please let me know.

**Comment by pekkaklarck** [comment]
Quick comments:

- I agree `keyword` is more explicit than `kw`, but if it is repeated huge amount of times then saving 5 characters per usage can be quite a lot. One idea of the JSON output would be making the output file smaller.
- Using Boolean values and not "yes" and "no" is definitely a good idea.
- In the end JSON output should be activated simply by using `--output output.json` which should then also disable writing the normal output.xml flle. It should work both with `robot` and `rebot`.
- This is such big change and RF 3.2 is so close that this can be earliest included to RF 3.3 (which may actually be RF 4.0).
- The same schema and lot of the code could be reused to create a tool that can convert current ouput.xml files to output.json (and possibly also the other way around). That would be a great way to make sure the schema works well as it could be fine tuned freely. Once it's taken into use by RF itself, then all changes are subject to backwards compatibility consideration.

**Comment by Lemonlemmings** [comment]
Okay, so I have updated my code. Robot will now output a .json file.
- I have changed the properties back to `kw` and `doc` to be consistent with how robot is elsewhere. I've updated the schema to reflect these changes as well as the underlying code.
- I've updated the JSON writer to not be dependent on requiring the entire output upfront (i.e: Moving away from a writer to the logger format). This allows robot to recognise if the output ends in .json and will output the results in JSON format. I've tested this against a couple of inputs and it looks good on the whole.
- I'm in the middle of writing a `JsonExecutionResultBuilder` so that rebot can accept .json files in the schema format and make them available to as an ExecutionResult. If I'm correct then this will allow rebot to accept a mix of JSON and XML to merge. Hopefully this will also satisfy your requirement for a tool to be able to convert between the XML and JSON (as rebot will do this).

Should I open a pull request so you can track the changes, as I'm also renaming some functions (so robot can remain agnostic to what output format is chosen).

**Comment by Lemonlemmings** [comment]
Another update, rebot is now able to accept the .json format as an input. This allows it to combine with either other XML files, or other JSON files. It can produce either XML output, or JSON output. I've tested it against a couple of examples so far and it seems to work fine.
- It needs a little tidying still. I know robot prefers classes for the element handlers (which I have no done on my implementation).
- It needs some unit tests and acceptance tests writing still.
- I haven't run this on python2 yet. My expectation is that it should work, but I haven't gotten round to installing python2 to check yet.

**Comment by pekkaklarck** [comment]
Sounds great! Quick comments from me again:

- A PR showing the code would be good. Doesn't need to be fully ready.
- Example of the JSON output would be interesting. Perhaps a [gist](https://gist.github.com/)?
- Have you done any performance measurements? I'm especially interested in the output.json processing speed.
- Could you please inform people on the #devel channel in our Slack about this? I'm sure others would be interested. This would require a PR so that they can test it.
- If changes are relatively small and benefits are big enough, we could possibly still get this into RF 3.2.

**Comment by pekkaklarck** [comment]
Ping @Muusssi!

**Comment by Lemonlemmings** [comment]
Unfortunately I haven't found the time to progress this forward more.
- I have opened a pull request: [pull request](https://github.com/robotframework/robotframework/pull/3475)
- I have made a gist with the output of what I've been testing: [JSON gist](https://gist.github.com/Lemonlemmings/40f137050f00d72613a84504091fd0ef)
- I am interested in the speeds, but I haven't gotten around to doing this yet. I'll have a look this weekend.
- I've not used Slack before, I'll look at how to do this at the weekend 😄 
- It would be nice to see if we could get it into RF 3.2

**Comment by SeleniumUser** [comment]
I like the idea to generate output in json format . i have similarly situation where i want output in json format so that i can upload test result in other external tracking system. May i request you to please update the RF3.2 release date.

**Comment by pekkaklarck** [comment]
RF 3.2 will be released very soon. This issue needs to wait for future versions.

**Comment by Lemonlemmings** [comment]
I've been looking at this again today, I've updated it so the unit tests pass, but I'm suffering from encoding issues with strings for acceptance testing.

**Comment by Lemonlemmings** [comment]
Sorry it's been so long since my last update. COVID-19 has only made me more busy (as I imagine it has many software engineers). I've come back and had a look at my code and realised it was left in a bit of a state (it would produce various different errors, and wasn't properly forming the models with the element handlers).

I've resolved these issues and I have been running some tests with rebot and found that I can convert freely between XML and JSON using rebot, and that the data trasformation between the two is consistent!! (Hooray) I have also ensured that the output from the JSON validates against the schema! (Hooray again).

I have also done some testing with timings, although these have been very limited tests. I was noticing, with a fairly basic robot file, that XML was marginally quicker.

XML: 0.033s
JSON: 0.036s

I don't know if this changes with larger robot files. I imagine this might change based off of the kind of hard disk you have available (I have an NVMe SSD) as the JSON implementation will store the results in memory and will write to disk when it's complete, rather than the XML implementation which looks like it writes to disk as it goes. This implementation is unfortunate as it makes the JSON implementation harder to do with the visitor pattern because you have to remember "where you are" between visits.

I'm increasingly finding that my implementation of the JSON is stable, but the other surrounding bits (like recognising the output file ends in ".json", and the addition of the "--json") don't quite work how I'd expect them to. So my next task is to ensure that the flows are working how I'd expect them to be. Afterwards I'll look to implement the other options that rebot supplies as it looks like some of the options around flattening are completed when the XML is loaded and is something I'll have to in the loading of the JSON. All in all, seems to be coming along nicely!

I have two questions. 

In my pull request, I'm getting some failures that I wouldn't expect to be getting based off of the code I've changed. Are there known issues with the existing set of acceptance tests, or have I done something wrong (such that I'd be getting failures with the help lines)?

I'm getting failures in Python2 acceptance tests because there are a whole load of string classes robot has for dealing with unicode strings in Python2 and this grinds against my implementation. For recognition of the output ending in ".json" I have implemented a line of code like so:
`output.lower().endswith(".json")`. This stops working with the custom string classes because they don't have these functions available. These custom string classes appear to only be in python2 implementations because I only get the error here, not in python3 acceptance tests. Is this something I need to care about because python2 is now deprecated, or should I go to the effort of finding a way around this problem?

**Comment by Lemonlemmings** [comment]
I've just run an example on my laptop (with a SATA SSD). This ia across 20 runs of the same robot script:

JSON = 3.6550284
XML = 3.6984686

My laptop has slower specs, but I think the notable difference is that it has a SATA SSD rather than an NVMe drive which benefits the performance of the JSON implementation.

This is the example robot file (simple): https://gist.github.com/Lemonlemmings/c6b33ea88e68327cc63afd5c6c793f26

Since it's been a while here is the outputted JSON from that run: https://gist.github.com/Lemonlemmings/cc92e26f3edd02ae40d700352ef63e77

**Comment by Murugan-Ramaswamy** [comment]
Thank you for adding this feature , and expecting it to be released soon as I got a requirement to have the output.xml converted to output.json, Actually I'm  trying to develop the code for this conversion and facing difficulties and when I googled for solution I ended up in this link and happy to see such feature is being added, Expecting the release soon and have to thank and Congratulate this team for the good effort, Great work guys..!

**Comment by skhomuti** [comment]
To increase the parsing speed - maybe try other parsers? https://github.com/ultrajson/ultrajson or https://github.com/ijl/orjson.
I think I can check out the different options on our large output.xml files.

**Comment by mkorpela** [comment]
Related https://en.wikipedia.org/wiki/JSON_streaming

**Comment by mkorpela** [comment]
I took a look at the json PR. There are several things to consider.

I would like that also the old issue with output is fixed. This goes back to the time of implementing the current log.html data format.
There are two problems in my opinion:
1. **Whole output must be read completely when processing it.** This means that there is a memory spike. Pabot can somewhat currently solve this for humongous projects by removing keywords for partial results.
2. **Output is broken when writing while file is stopped in the middle** Data is not directly usable without fixing it.

I propose that the output.json when done would be [Line-delimited JSON](https://en.wikipedia.org/wiki/JSON_streaming#Line-delimited_JSON).
In a format that can be also partially processed even when writing has aborted.

Now when I tested the output.json PR everything is written in the same line
and has the same tree like structure as the original output.xml.
This means that reading problems are not handled.

In the best case scenario writing log.html from output.json would not have a memory spike.
Memory spike means that data processing requires RAM that might not be available.
Data would just flow from streaming json that is only partially read and transformed to model objects and from that directly to the format that log.html and report.html use.

Output.xml is also very verbose and replicates same elements.
For example keyword documentation is logged every time that a keyword is executed.
status elements also eat alot of space as one can check with this gist https://gist.github.com/mkorpela/0a38a0442e601766568eb7823297d44c

**Comment by Lemonlemmings** [comment]
Thanks for your comment! Sorry it's been a while since I've posted anything. COVID has made things quite mad (I'm sure it has for everyone). I've posted up in the Slack, and I'm hoping to get this release into RF4.1.

In response to your points:

I agree loading a large project directly into memory isn't ideal. To save me a lot of time rewriting something from scratch I have found [this library](https://pypi.org/project/jsonslicer/). I haven't done an implementation with it yet, but I think this would resolve your problem, it would only load something in when needed (if you do the loading correctly). The project mentions loading large JSON files very quickly, so that should be resolve the issue. **One minor thing...** that paticular project only supports Python3, and does not support Python2.7 (which RF4 will continue support for). With that in mind, I propose that I do an implementation with that library which will allow for two situations:
1. If Python3 is being used (i.e: That library is installed) then it will use the nicer streaming option
2. If Python2.7 is being used, and the JSON output is being activated it will simply load the output into memory (unfortunate, but this would go away with RF5 and Python2 support being dropped)

I have (slightly) resolved your problem with the output being broken if execution is stopped midway by using [this lovely library](https://pypi.org/project/jsonstreams/). This means that for very large projects this JSON implementation won't consume lots of memory (like the implementation you tested).

Things that are leftover:

1. I need to finish my implementation for reading from the JSON (see above).
2. This needs testing in a range of scenarios, I think as a base run it would be good to run the acceptance tests with the JSON output (see how it compared)
3. I need to finish the little off the trimmings I have added elsewhere. Currently it will recognise if the output file ends in JSON, but the JSON flag itself isn't working how I'd expect it too.
4. Tests needs implementing in the unit and acceptance tests.

**Comment by pekkaklarck** [comment]
Quick comments:
1. Having this in RF 4.1 would be nice. 
2. I'm fine this feature being Python 3 only even if the release would still support Python 2.
3. RF doesn't currently have any mandatory external dependency and I don't want that to change. Some features like YAML variable files only work if you install an external module, and we could use the same approach also with output.json.
4. Do you already have a JSON structure planned?

**Comment by Lemonlemmings** [comment]
Hi, thanks for your comments.

1. I agree, I would like to see this in RF 4.1 as well!
2. I think I can make it so it's both Python 2 and Python 3 (one would be efficient, and the other not)
3. With the above comment in mind, I'm going to propose that I do the work to do JSON streaming using an external dependency (which is Python 3 only), and that I make the format work if you simply load all of the JSON into memory instead. Essentially meaning that if the library is imported that you don't have memory spikes when parsing JSON
4. I do already have a structure planned. It fairly closely matches the XML format. I have produced a schema (and the work I have done matches that schema) - https://github.com/Lemonlemmings/robotframework/blob/master/doc/schema/robot-10.schema.json (This schema is updated for the RF 4.0 changes)

**Comment by Lemonlemmings** [comment]
I've been doing some additional work this weekend.

- I've fixed the flags not working in the way I expect. The robot module has `--json` which is a flag which marks the output as JSON. It will also rename the default output from `output.xml` to `output.json`. For rebot `--json` is a path to the output. The only situation I see as being awkward (as it stands) is that rebot has no way of recognising if the input is a JSON output format if the filename doesn't end in `.json`. I'm not quite sure how to approach this problem without completely changing how rebot accepts arguments (which I was trying to avoid doing).
- I have marked the library I use to do the JSON as being optional. I've noticed that it's MIT, and since it's a single file could easily be integrated into RF (to avoid the external dependency issue). I'm not sure what the approach would be here, but I'm happy to hear thoughts.
- I have discovered that the libary I had found for using JSON slicing is fairly difficult to get installed. I think, for now, that the JSON input will cause a memory spike when it's being read in, and there's not a lot that can be done to avoid that without writing a JSON streamer (which I feel is out of scope). I guess if this is an issue for users, then they should be using the XML output (which has had a lot more optimisation than this code has).

In regards to my TODO list I have the following items left:

- I need to implement flattening of keywords still. I'm a little confused as to how that would work exactly. I wouldn't mind a brief explanation. I think I can easily apply it to the JSON format with some recursive functions.
- In a similar vein to the above, I have noticed the `_omit_keywords` function on the `XmlExecutionResultBuilder` class (which I think is normally `ExecutionResultBuilder`). It seems to only omit keywords, but I was wondering if this was purposeful, as this doesn't include removal of `FOR`, `ITER`, `IF`, or `BRANCH`. Is this is purposeful, then I can implement something similar. I felt it was odd to have a format which only removes keyword data, and not the surrounding structure.
- Write acceptance tests for JSON output format. My current stance is to copy the existing tests for the XML format, and make them apply to the JSON format instead. I was thinking that it would be good to do some kind of back and forth with rebot, as rebot allows the JSON format to be converted into XML and back again (so there should be some tests to ensure that the format always produces the same results).

**Comment by pekkaklarck** [comment]
Issue #3902 proposes adding support for serializing executable `TestSuite´ objects into JSON and being able to recreate suites based on this data. The executable suite and the result suite are similar, they have a common base, and they should preferably use same JSON serialization approach as well. As you can read in comments of #3902, I [propose supporting splitting a suite into multiple objects](https://github.com/robotframework/robotframework/issues/3902#issuecomment-805280596) using [line-delimited JSON](https://en.wikipedia.org/wiki/JSON_streaming#Line-delimited_JSON) that @mkorpela mentioned in an [earlier comment](https://github.com/robotframework/robotframework/issues/3423#issuecomment-743253796) to this issue.

Unfortunately splitting a suite into multiple objects as proposed in #3902 is problematic during execution. The problem is that when, for example, a suite starts we don't yet have all information (end time and status aren't known, child suites and tests may be dynamically modified, ...) and thus cannot write the final suite object even with placeholders for child elements. At least these possibilities could considered to avoid this problem:
1. Write a "one big object" during execution instead of splitting a suite. This approach has drawbacks:
    - All benefits of having multiple objects are lost.
    - If we also want to support splitting objects, we either need to write the "one big object" into a single line or needed to somehow detect during parsing when a multiline file has multiple objects and when just one object split to multiple lines.
2. Write the info we have and when the suite (or test or keyword) rewrite it. This is complicated because we need to move around the file and especially because new entries would generally be bigger and thus overwrite other data. Always rewriting the whole file would be inefficient and reserving extra space after each entry to allow expanding would be complicated.
3. When a suite (or test or keyword) ends, write a new entry with new and updated information. We'd just need a way to link the new and old entries. We could, for example, store `id(obj)` as and `_id` value to these objects just for this purpose.

**Comment by pekkaklarck** [comment]
Good catch @Lemonlemmings that `ExecutionResultBuilder._omit_keywords` doesn't omit FORs and IFs. They definitely should but luckily this bug doesn't affect functionality, only performance, and because these structures ought to be rare on the test case level not many is affected in general. Submitted #3903 about that anyway.

**Comment by Lemonlemmings** [comment]
Hi,

Thanks for your comment.

In regards to using line delimited JSON there are also drawbacks. As you have noted, there would need to be some ID correlation between lines that you're writing in the JSON. It feels like you want to use JSON as a concatenative protocol when  JSON is a non-concatenative protocol. The XML execution format already fills this niche (and has been very well optimised for avoid memory spikes, and streaming).

I have managed a lot of the problems with the "one big object" using the[ jsonstreams](https://github.com/dcbaker/jsonstreams/blob/main/jsonstreams/__init__.py) module. This is a single python file which will let you "open" and "close" JSON while writing directly to disk. This means you don't have to worry about storing the JSON object you're building in memory (it is written directly to disk). I get quite unclear about licencing, but I believe that as the project is released under MIT (and also tests successfully against Python2 and Python3) and that it would be suitable for inclusion in the `utils` package in Robot. This would allow for "native" execution of the JSON execution output. My implementation also has an interface between the execution output and the jsonstreams library, which could be interchanged for an in-memory library, or perhaps even a HTTP library (which would be useful for the pabot rewrite as you can send objects back to a central server when they're closed).

The issue I face with the JSON is that you cannot really get around the fact that you have to load an entire JSON object into memory if you want to process it into the log or report files. This is an advantage that the multiple objects has over the "one big object" approach.

For my personal circumstances I have been using the JSON output format for roughly the past year (ever since I finished the original code). I run robot script in different configurations en-masse, and then pipe the one big JSON result (with only the test case JSON data) into an ELK stack. This lets me do analytics over the suites I have executed, and compare them to one another nicely. I would think that splitting the JSON into multiple objects would make this approach very difficult.

**Comment by Lemonlemmings** [comment]
Upon further reading of your suggest, I realise I got the wrong impression of what you were looking at. I got the impression that you were looking to only use a line delimited JSON option, and looking to drop any potential "one big object" JSON.

What you're suggesting with your multiple objects is something we can achieve using the interface I have currently implemented. The "setup", "teardown", "tests", "root suite" and "suites" already have distinct code paths in my implementation, so potentially starting a new line would be possible when looking to execute any of these items. Are you proposing that when you're executing with JSON that you could also add an optional flag of `--split` that not only would execute with the JSON output, but would cause it to split on your objects for a new-line delimited JSON? If so, then this feels achievable alongside the "one big result" format we already have implemented. Having said that though, the jsonstreams module alone would not be enough to achieve what you want.

**Comment by Lemonlemmings** [comment]
If you are confident that you'll always finish the robot suite, then you could accept that your multiple objects would be written to disk in the wrong order, and then have some code that reorders them after the suite has finished executing. You would write them in the order than the objects "close" or "end". Then you use the ID's you assign to reorder them after the execution has finished. It feels a little cheap, and wouldn't provide very useful information in the event of failure (although this is already the case for the jsonstreams which will leave the file without the necessary closing braces in the event of failure (although I think this is a quick fix to capture exceptions and ensure we close the object before raising the exception again.

**Comment by pekkaklarck** [comment]
I haven't thought about all possible details related to the possible new JSON serialization format so cannot really answer the questions you have regarding to it. In that usage writing JSON isn't a big problem, the hard part is reading it efficiently. That's why I was thinking to allow splitting JSON to multiple objects so that we didn't need to read the whole JSON into memory in one go. I added a comment to #3902 about how to do that in practice and how we could remove duplication also otherwise:
https://github.com/robotframework/robotframework/issues/3902#issuecomment-808405017

**Comment by pekkaklarck** [comment]
Using JSON as a serialization format for data is considerably simpler problem than using JSON for results, especially because we should be writing results to file during execution when we don't have all information available yet. It might thus be better to concentrate on the former first. We should nevertheless keep results in mind to avoid solutions that were not compatible with them if at all possible.

**Comment by pekkaklarck** [comment]
This is too big issue for RF 4.1 that is planned to have only smallish enhancements before the bigger RF 5.0 release. I move this into RF 5.0 scope instead. Then we don't need to care about Python 2 either making this implementation easier.

**Comment by soundarameena** [comment]
Hello . So after reading above comments came to know there was some progress in this json output . May I please know how can it be done or is there some direct way to do get this json output from robotframework instead of xml output? We in our company need this json output to feed to Zephyr to get test reports. Can anyone help me?

**Comment by pekkaklarck** [comment]
No progress lately and this isn't too high in the priority list at the moment. Would be a nice feature but implementing it properly is a rather big task. What you can do now is converting the output.xml to JSON yourself after execution.

**Comment by soundarameena** [comment]
Hello

Thank you so much for getting back to me. As you said I tried to convert an
output.xml to json and was successful with it.

But now the issue is that we need cucumber format json which is only
accepted by zephyr. Is there any solution to convert output.xml into a
cucumber format json?

Any ideas will be really helpful in progressing our work.


Thanks
Nithya

**Comment by pekkaklarck** [comment]
You'd need to study how the Cucumber JSON output looks like and generate JSON with same structure based on output.xml. Notice that if/when output.json is added to Robot, it is very unlikely that it would be compatible with Cucumber. Finally, there could be better approaches for Zephyr integration but that's not really the topic of this issue. I recommend you ask about this from other users on our Slack or Forum. You can find both via robotframework.org.

**Comment by pekkaklarck** [comment]
The plan is to add JSON serialization support to the result model in RF 7 (#4847). Writing output.json as part of execution is outside the scope of that issue, that's what this issue is all about, but it will define JSON format. The plan is also that Rebot could produce output.json after execution. That would mean that you can do

```
robot tests.robot
rebot --output output.json output.xml
rm output.xml
```
to get output.json instead of output.xml.

It would obviously be more convenient to be able to generate output.json already during execution, but that's a lot more complicated because it would need a separate JSON writer that can write information in pieces as soon as we have the info available. It would also need to support flattening and other such features that our existing XML writer does. I hope we get that done at some point, but that's too big task for RF 7.

**Comment by pekkaklarck** [comment]
JSON serialization support has now been added to Robot's result model (#4847) and also Rebot supports it so that the example in the above comment works. This is great, but obviously not as convenient than getting output.json directly as part of the execution.

We now already have the JSON format specified, so there's less to be designed when we return to this. Some of the key things that need to be done are listed below:
1. `JsonLogger` class that can write information to disk when suites, tests, keywords, etc. start and end. We probably need to create it ourselves similarly as we have created our `XmlLogger`. These classes should probably have same API so that it's easy to switch the logger without changing other code. 
2. `XmlLogger` has a pretty ok API, but there probably are things that need to be fixed in it before adding another class that mimics it.
3. We need to decide on the top level output.json format. Current output.json created by Rebot only has suite information, but we need to add also statistics and errors there similarly as we have in output.xml. Probably it's enough to add a new root object with keys `suite`, `statistics` and `errors`. Current methods processing output.json files need to support both the "full" output.json files and files only containing suite information.
4. `Result` object should probably get its own `to/from_xml` and `to/from_json` methods for conveniently serializing it to different formats. The old `save` method should be preserved, though.

**Comment by pekkaklarck** [comment]
Let's see can we get this done already in RF 7.2. As mentioned above, the result model can nowadays be serialized to JSON and Rebot also supports JSON outputs, so we don't need to discuss about the suite structure anymore. We need to define the higher level result structure, but that's covered by #5160 that's already in RF 7.2 scope. The remaining task related to this issue is creating a logger that can write JSON incrementally. Also that has challenges like taking flattening into account, but I don't expect it to be too complicated.

**Comment by pekkaklarck** [comment]
This is now mostly ready and thoroughly unit tested. Things to do:
- Acceptance tests including schema validation.
- Handling keywords executed by listeners in random places.
- Documentation.

Implementation turned out to be pretty complicated due to JSON syntax annoyances (trailing commas not being allowed required keeping state) and Robot's internal logging not supporting this style too well (for example, when a test starts, we need to make sure to add `"tests":[` if it isn't added already and thus we need to keep more state). I'm happy how the code looks in the end, but getting it to this state took some time.

**Comment by pekkaklarck** [comment]
After the latest changes this feature is now done and ready to be released as part of RF 7.2 beta 1. Implementing this properly was a bigger task than I anticipated, so hopefully the functionality is useful.

**Comment by pegna12** [comment]
Hello @pekkaklarck,
This sounds very useful indeed. Do you know if there is a plan to add support for output.json in the Jenkins RF plugin?
